src/framework/uicomponents/view/popupview.cpp

```diff
@@ -405,14 +406,144 @@ void PopupView::setRet(QVariantMap ret)
     emit retChanged(m_ret);
 }
 
+void PopupView::setOpensUpward(bool opensUpward)
+{
+    if (m_opensUpward == opensUpward) {
+        return;
+    }
+
+    m_opensUpward = opensUpward;
+    emit opensUpwardChanged(m_opensUpward);
+}
+
+void PopupView::setArrowX(int arrowX)
+{
+    if (m_arrowX == arrowX) {
+        return;
+    }
+
+    m_arrowX = arrowX;
+    emit arrowXChanged(m_arrowX);
+}
+
+void PopupView::setCascadeAlign(Qt::AlignmentFlag cascadeAlign)
+{
+    if (m_cascadeAlign == cascadeAlign) {
+        return;
+    }
+
+    m_cascadeAlign = cascadeAlign;
+    emit cascadeAlignChanged(m_cascadeAlign);
+}
+
 QVariantMap PopupView::ret() const
 {
     return m_ret;
 }
 
+bool PopupView::opensUpward() const
+{
+    return m_opensUpward;
+}
+
+int PopupView::arrowX() const
+{
+    return m_arrowX;
+}
+
+Qt::AlignmentFlag PopupView::cascadeAlign() const
+{
+    return m_cascadeAlign;
+}
+
 void PopupView::setErrCode(Ret::Code code)
 {
     QVariantMap ret;
     ret["errcode"] = static_cast<int>(code);
     setRet(ret);
 }
+
+QRect PopupView::currentScreenGeometry() const
+{
+    QScreen* currentScreen = QGuiApplication::screenAt(m_globalPos.toPoint());
+    if (!currentScreen) {
+        return QRect();
+    }
+
+    return currentScreen->geometry();

```

Comment:
How about using <code>**available**Geometry()</code> here instead, to avoid collision with the system's menu bar and task bar etc?

Reply:
thanks! all right, fixed